presto-hive-common/src/main/java/com/facebook/presto/hive/util/DirectoryLister.java

```diff
@@ -11,18 +11,16 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.facebook.presto.hive;
+package com.facebook.presto.hive.util;
 
+import com.facebook.presto.hive.HiveFileInfo;
 import com.facebook.presto.hive.metastore.Table;
-import org.apache.hadoop.fs.FileSystem;
+import com.google.common.base.Supplier;
 import org.apache.hadoop.fs.Path;
-import org.apache.hadoop.fs.PathFilter;
 
 import java.util.Iterator;
 
-import static com.facebook.presto.hive.util.HiveFileIterator.NestedDirectoryPolicy;
-
 public interface DirectoryLister
 {
-    Iterator<HiveFileInfo> list(FileSystem fileSystem, Table table, Path path, NamenodeStats namenodeStats, NestedDirectoryPolicy nestedDirectoryPolicy, PathFilter pathFilter);
+    Iterator<HiveFileInfo> list(Path path, Table table, Supplier<Iterator<HiveFileInfo>> hiveFileInfoSupplier);
```

Comment:
This is also a weird interface where the input is a iterator supplier and the output is the iterator.... I think both `CachingDirectoryLister` and `TimeoutDirectoryLister` should take a 'delegate" in their constructors rather than providing a supplier here. The supplier passed in from `PrestoExtendedFileSystemWrapper` or any other FS should create their own generic `DirectoryLister`.

The old logic is exactly like that, isn't it?

Reply:
Yes the old logic was that. The reason this is different because the initiator here is the FS rather DirectoryLister. 
The DirectoryLister is a utility which takes an iterator of HiveFileInfo and then depending on its own logic uses this input iterator supplier conditionally and return to the consumer.

Reply:
The other way to clean this would be to have two list methods in the ExtendedFileSystem, or single one with a parameter to signal if the DirectoryLister should be used or not. Then it would look like following:
ExtendedFS::listwithDL => DirectoryLister::list => ExtendedFS::listwithoutDL